File: app/frontend/src/api/email.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import apiClient from './client'

export interface SmtpConfig {
  id: string
  host: string
  port: number
  username: string
  use_tls: boolean
  from_email: string
  from_name: string
  is_active: boolean
}

export interface EmailTemplate {
  id: string
  name: string
  slug: string
  subject: string
  html_body: string
  available_vars: Record<string, string> | null
  is_system: boolean
  created_at: string
  updated_at: string
}

export interface EmailWorkflow {
  id: string
  name: string
  is_active: boolean
  trigger_type: string
  trigger_config: Record<string, any> | null
  condition_config: Record<string, any> | null
  template_id: string
  recipient_type: string
  recipient_users: string[] | null
  send_teams: boolean
  teams_webhook_id: string | null
  last_run_at: string | null
  created_at: string
  updated_at: string
}

export interface EmailLog {
  id: string
  workflow_id: string | null
  template_id: string | null
  recipient_id: string | null
  to_email: string
  subject: string
  status: 'pending' | 'sent' | 'failed'
  error_message: string | null
  sent_at: string | null
  created_at: string
}

export interface EmailLogListResponse {
  items: EmailLog[]
  total: number
  skip: number
  limit: number
}

export interface TeamsWebhook {
  id: string
  name: string
  is_active: boolean
  created_at: string
}

const emailKeys = {
  all: ['email'] as const,
  smtp: () => [...emailKeys.all, 'smtp'] as const,
  templates: () => [...emailKeys.all, 'templates'] as const,
  workflows: () => [...emailKeys.all, 'workflows'] as const,
  logs: (params?: object) => [...emailKeys.all, 'logs', params] as const,
  teamsWebhooks: () => [...emailKeys.all, 'teams-webhooks'] as const,
}

// ─── SMTP ────────────────────────────────────────────────────────────────────

export function useSmtpConfigs() {
  return useQuery({
    queryKey: emailKeys.smtp(),
    queryFn: () => apiClient.get<SmtpConfig[]>('/email/smtp').then((r) => r.data),
  })
}

export function useCreateSmtpConfig() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: Omit<SmtpConfig, 'id' | 'is_active'> & { password: string }) =>
      apiClient.post<SmtpConfig>('/email/smtp', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.smtp() }),
  })
}

export function useUpdateSmtpConfig() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...data }: { id: string; password?: string; is_active?: boolean; [key: string]: any }) =>
      apiClient.patch<SmtpConfig>(`/email/smtp/${id}`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.smtp() }),
  })
}

export function useTestSmtpConfig() {
  return useMutation({
    mutationFn: (id: string) =>
      apiClient.post<{ success: boolean; message: string }>(`/email/smtp/${id}/test`).then((r) => r.data),
  })
}

// ─── Templates ────────────────────────────────────────────────────────────────

export function useEmailTemplates() {
  return useQuery({
    queryKey: emailKeys.templates(),
    queryFn: () => apiClient.get<EmailTemplate[]>('/email/templates').then((r) => r.data),
  })
}

export function useCreateEmailTemplate() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: { name: string; slug: string; subject: string; html_body: string; available_vars?: object }) =>
      apiClient.post<EmailTemplate>('/email/templates', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.templates() }),
  })
}

export function useUpdateEmailTemplate() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...data }: { id: string; name?: string; subject?: string; html_body?: string }) =>
      apiClient.patch<EmailTemplate>(`/email/templates/${id}`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.templates() }),
  })
}

export function useDeleteEmailTemplate() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => apiClient.delete(`/email/templates/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.templates() }),
  })
}

export function usePreviewEmailTemplate() {
  return useMutation({
    mutationFn: ({ id, variables }: { id: string; variables: object }) =>
      apiClient.post<{ html: string }>(`/email/templates/${id}/preview`, { variables }).then((r) => r.data),
  })
}

// ─── Workflows ────────────────────────────────────────────────────────────────

export function useEmailWorkflows() {
  return useQuery({
    queryKey: emailKeys.workflows(),
    queryFn: () => apiClient.get<EmailWorkflow[]>('/email/workflows').then((r) => r.data),
  })
}

export function useCreateEmailWorkflow() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: any) => apiClient.post<EmailWorkflow>('/email/workflows', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.workflows() }),
  })
}

export function useUpdateEmailWorkflow() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...data }: { id: string; [key: string]: any }) =>
      apiClient.patch<EmailWorkflow>(`/email/workflows/${id}`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.workflows() }),
  })
}

export function useToggleEmailWorkflow() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => apiClient.patch<EmailWorkflow>(`/email/workflows/${id}/toggle`).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.workflows() }),
  })
}

export function useDeleteEmailWorkflow() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => apiClient.delete(`/email/workflows/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.workflows() }),
  })
}

// ─── Logs ─────────────────────────────────────────────────────────────────────

export function useEmailLogs(params?: { status?: string; skip?: number; limit?: number }) {
  return useQuery({
    queryKey: emailKeys.logs(params),
    queryFn: () => apiClient.get<EmailLogListResponse>('/email/logs', { params }).then((r) => r.data),
  })
}

// ─── Teams Webhooks ───────────────────────────────────────────────────────────

export function useTeamsWebhooks() {
  return useQuery({
    queryKey: emailKeys.teamsWebhooks(),
    queryFn: () => apiClient.get<TeamsWebhook[]>('/email/teams-webhooks').then((r) => r.data),
  })
}

export function useCreateTeamsWebhook() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: { name: string; webhook_url: string }) =>
      apiClient.post<TeamsWebhook>('/email/teams-webhooks', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.teamsWebhooks() }),
  })
}

export function useDeleteTeamsWebhook() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => apiClient.delete(`/email/teams-webhooks/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: emailKeys.teamsWebhooks() }),
  })
}

export function useTestTeamsWebhook() {
  return useMutation({
    mutationFn: (id: string) => apiClient.post<{ success: boolean }>(`/email/teams-webhooks/${id}/test`).then((r) => r.data),
  })
}

File: app/backend/app/routers/email.py
import uuid
from typing import Annotated, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.user import User
from app.schemas.email import (
    SmtpConfigCreate, SmtpConfigUpdate, SmtpConfigResponse,
    EmailTemplateCreate, EmailTemplateUpdate, EmailTemplateResponse, EmailTemplatePreviewRequest,
    EmailWorkflowCreate, EmailWorkflowUpdate, EmailWorkflowResponse,
    EmailLogResponse, EmailLogListResponse,
    TeamsWebhookCreate, TeamsWebhookUpdate, TeamsWebhookResponse,
)
from app.schemas.auth import MessageResponse
from app.services.email_service import EmailService
from app.services.teams_service import TeamsService
from app.core.dependencies import get_current_user, require_superadmin, require_manager_or_above

router = APIRouter(prefix="/email", tags=["email"])


# ─── SMTP ─────────────────────────────────────────────────────────────────────

@router.get("/smtp", response_model=list[SmtpConfigResponse])
async def list_smtp(
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.list_smtp_configs()


@router.post("/smtp", response_model=SmtpConfigResponse, status_code=201)
async def create_smtp(
    body: SmtpConfigCreate,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.create_smtp_config(
        host=body.host, port=body.port, username=body.username,
        password=body.password, use_tls=body.use_tls,
        from_email=body.from_email, from_name=body.from_name,
    )


@router.patch("/smtp/{config_id}", response_model=SmtpConfigResponse)
async def update_smtp(
    config_id: uuid.UUID,
    body: SmtpConfigUpdate,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.update_smtp_config(
        config_id, host=body.host, port=body.port, username=body.username,
        password=body.password, use_tls=body.use_tls,
        from_email=body.from_email, from_name=body.from_name, is_active=body.is_active,
    )


@router.post("/smtp/{config_id}/test", response_model=dict)
async def test_smtp(
    config_id: uuid.UUID,
    current_user: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Send a test email via the given SMTP configuration."""
    svc = EmailService(db)
    result = await svc.test_smtp_config(config_id, to_email=current_user.email)
    return result


# ─── Templates ────────────────────────────────────────────────────────────────

@router.get("/templates", response_model=list[EmailTemplateResponse])
async def list_templates(
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.list_templates()


@router.post("/templates", response_model=EmailTemplateResponse, status_code=201)
async def create_template(
    body: EmailTemplateCreate,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.create_template(
        name=body.name, slug=body.slug, subject=body.subject,
        html_body=body.html_body, available_vars=body.available_vars,
    )


@router.patch("/templates/{template_id}", response_model=EmailTemplateResponse)
async def update_template(
    template_id: uuid.UUID,
    body: EmailTemplateUpdate,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.update_template(
        template_id, name=body.name, subject=body.subject,
        html_body=body.html_body, available_vars=body.available_vars,
    )


@router.delete("/templates/{template_id}", response_model=MessageResponse)
async def delete_template(
    template_id: uuid.UUID,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    await svc.delete_template(template_id)
    return {"message": "Şablon silindi."}


@router.post("/templates/{template_id}/preview")
async def preview_template(
    template_id: uuid.UUID,
    body: EmailTemplatePreviewRequest,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    html = await svc.preview_template(template_id, body.variables)
    return {"html": html}


# ─── Workflows ────────────────────────────────────────────────────────────────

@router.get("/workflows", response_model=list[EmailWorkflowResponse])
async def list_workflows(
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.list_workflows()


@router.post("/workflows", response_model=EmailWorkflowResponse, status_code=201)
async def create_workflow(
    body: EmailWorkflowCreate,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.create_workflow(
        name=body.name,
        trigger_type=body.trigger_type,
        template_id=body.template_id,
        recipient_type=body.recipient_type,
        trigger_config=body.trigger_config,
        condition_config=body.condition_config,
        recipient_users=body.recipient_users,
        send_teams=body.send_teams,
        teams_webhook_id=body.teams_webhook_id,
    )


@router.patch("/workflows/{workflow_id}", response_model=EmailWorkflowResponse)
async def update_workflow(
    workflow_id: uuid.UUID,
    body: EmailWorkflowUpdate,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.update_workflow(
        workflow_id,
        name=body.name,
        trigger_config=body.trigger_config,
        condition_config=body.condition_config,
        template_id=body.template_id,
        recipient_type=body.recipient_type,
        recipient_users=body.recipient_users,
        send_teams=body.send_teams,
        teams_webhook_id=body.teams_webhook_id,
        is_active=body.is_active,
    )


@router.patch("/workflows/{workflow_id}/toggle", response_model=EmailWorkflowResponse)
async def toggle_workflow(
    workflow_id: uuid.UUID,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    return await svc.toggle_workflow(workflow_id)


@router.delete("/workflows/{workflow_id}", response_model=MessageResponse)
async def delete_workflow(
    workflow_id: uuid.UUID,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = EmailService(db)
    await svc.delete_workflow(workflow_id)
    return {"message": "İş akışı silindi."}


@router.post("/workflows/{workflow_id}/test-run", response_model=MessageResponse)
async def test_run_workflow(
    workflow_id: uuid.UUID,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    from app.tasks.email_tasks import evaluate_scheduled_workflows
    evaluate_scheduled_workflows.delay()
    return {"message": "İş akışı kuyruğa alındı."}


# ─── Logs ─────────────────────────────────────────────────────────────────────

@router.get("/logs", response_model=EmailLogListResponse)
async def list_logs(
    _: Annotated[User, Depends(require_manager_or_above)],
    db: Annotated[AsyncSession, Depends(get_db)],
    status: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
):
    svc = EmailService(db)
    items, total = await svc.list_logs(skip=skip, limit=limit, status=status)
    return {"items": items, "total": total, "skip": skip, "limit": limit}


# ─── Teams Webhooks ───────────────────────────────────────────────────────────

@router.get("/teams-webhooks", response_model=list[TeamsWebhookResponse])
async def list_teams_webhooks(
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = TeamsService(db)
    return await svc.list_webhooks()


@router.post("/teams-webhooks", response_model=TeamsWebhookResponse, status_code=201)
async def create_teams_webhook(
    body: TeamsWebhookCreate,
    current_user: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = TeamsService(db)
    return await svc.create_webhook(body.name, body.webhook_url, current_user.id)


@router.patch("/teams-webhooks/{webhook_id}", response_model=TeamsWebhookResponse)
async def update_teams_webhook(
    webhook_id: uuid.UUID,
    body: TeamsWebhookUpdate,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = TeamsService(db)
    return await svc.update_webhook(webhook_id, body.name, body.webhook_url, body.is_active)


@router.delete("/teams-webhooks/{webhook_id}", response_model=MessageResponse)
async def delete_teams_webhook(
    webhook_id: uuid.UUID,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = TeamsService(db)
    await svc.delete_webhook(webhook_id)
    return {"message": "Teams webhook silindi."}


@router.post("/teams-webhooks/{webhook_id}/test", response_model=dict)
async def test_teams_webhook(
    webhook_id: uuid.UUID,
    _: Annotated[User, Depends(require_superadmin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    svc = TeamsService(db)
    return await svc.test_webhook(webhook_id)
